presto-main/src/main/java/com/facebook/presto/metadata/ThriftRemoteNodeState.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.drift.client.DriftClient;
+import com.facebook.presto.server.thrift.ThriftServerInfoClient;
+import com.facebook.presto.spi.NodeState;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.Duration;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.units.Duration.nanosSince;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class ThriftRemoteNodeState
+        implements RemoteNodeState
+{
+    private final ThriftServerInfoClient thriftClient;
+    private final AtomicReference<Optional<NodeState>> nodeState = new AtomicReference<>(Optional.empty());
```

Comment:
Given that we just get/set this, it should be `volatile` and the atomic isn't necessary

Reply:
Kinda still leaning towards using `AtomicReference` in case someone might change the behavior in the future and miss the part... WDYT?

Reply:
Don't feel strongly, but if we were to adopt this handy checklist it would come up 😄 

https://github.com/code-review-checklists/java-concurrency#redundant-atomics